packages/test/end-to-end-tests/src/test/codePropsal.spec.ts

```diff
@@ -0,0 +1,148 @@
+             deltaConnectionServer, urlResolver);
+        return createAndAttachContainer(documentId, codeDetails, loader, urlResolver) as any as Container;
+    }
+
+    async function loadContainer(factoryEntries: Iterable<[string, IChannelFactory]>): Promise<Container> {
+        const factory = new TestFluidObjectFactory(factoryEntries);
+        const urlResolver = new LocalResolver();
+        const loader: ILoader = createLocalLoader(
+            [[codeDetails, factory],[codeDetails2,factory]],
+             deltaConnectionServer, urlResolver);
+        return loader.resolve({ url: documentLoadUrl }) as any as Container;
+    }
+
+    let container1: Container;
+    let container2: Container;
+    beforeEach(async () => {
+        deltaConnectionServer = LocalDeltaConnectionServer.create();
+
+        // Create a Container for the first client.
+        container1 = await createContainer([["map", SharedMap.getFactory()]]);
+
+        opProcessingController = new OpProcessingController(deltaConnectionServer);
+        opProcessingController.addDeltaManagers(container1.deltaManager);
+
+        await opProcessingController.process();
+
+        // Load the Container that was created by the first client.
+        container2 = await loadContainer([["map", SharedMap.getFactory()]]);
+        opProcessingController.addDeltaManagers(container1.deltaManager);
+
+        const quorum1 = container1.getQuorum();
+        const quorum2 = container2.getQuorum();
+
+        assert.deepStrictEqual(
+            quorum1.get("code"),
+            codeDetails,
+            "Code proposal in container1 doesn't match");
+
+        assert.deepStrictEqual(
+            quorum2.get("code"),
+            codeDetails,
+            "Code proposal in container2 doesn't match");
+
+        const dataObject1 = await requestFluidObject<ITestFluidObject>(container1, "default");
+        const map1 = await dataObject1.getSharedObject<ISharedMap>("map");
+
+        // BUG BUG quorum.propose doesn't handle readonly, so make sure connection is write
+        while (container1.deltaManager.connectionMode === "read" || !container1.connected) {
+            map1.set("foo","bar");
+            await Promise.all([
+                new Promise((resolve) => container1.connected ? resolve() : container1.once("connect", resolve)),
+                opProcessingController.process(),
+            ]);
+        }
+    });
+
+    it("Code Proposal", async () => {
+        container1.on("contextChanged",(c)=>{
+            assert.deepStrictEqual(
+                c,
+                codeDetails2,
+                "container1 context should be update");
+        });
+
+        container2.on("contextChanged",(c)=>{
+            assert.deepStrictEqual(
+                c,
+                codeDetails2,
+                "container2 context should be update");
+        });
+
+        await Promise.all([
+            container1.getQuorum().propose("code", codeDetails2),
```

Comment:
I think it should propose something other than `codeDetails2`, or we should have more resetting happening between tests. Since it is the exact same instance of `pkg`, I think there's a chance that even with accepting the code proposal, contextChanged event would not be raised, because the code in the Container event handler ignores the accepted code proposal if it is the same as the current one.